## Data preparation with PANDAS

In [114]:
import pandas as pd


# Power plant conditions
p_conditions = pd.DataFrame({"Power plant":           ["Ahlen", "Fjället",  "Forsen",   "Kärret"],
                           "Initial reservoir level": [5800e6,  1000e6,     20e6,       13e6    ],
                           "Maximum reservoir level": [7160e6,  1675e6,     27e6,       13e6    ],
                           "Minimum reservoir level": [5800e6,  1000e6,     10e6,       6e6     ],
                           "Discharge capacity":      [540,     135,        975,        680     ],
                           "Power conversion":        [0.52,    1.17,       0.29,       0.05    ],
                           "Maximum spillage":        [820,     930,        360,        400     ],
                           "Local flow":              [177,     28,         8,          29      ],
                          })

# Time conditions
t_conditions = pd.DataFrame({"Time": range(1,13),
                             "Cost": [45, 55, 95, 80, 140, 150, 80, 70, 130, 0, 0, 0]})

# Flow conditions (Connections between the power plants)
f_conditions = pd.DataFrame({"From": ["Ahlen",  "Fjället",  "Forsen"],
                             "To":   ["Forsen", "Forsen",   "Kärret"],
                             "Time": [3,        2,          2]
                            })


## GAMSPy FTW

### Initializing all our Parameters and Varaibales etc

In [115]:

from gamspy import Container, Set, Variable, Parameter, Equation, Sum, Model, Sense, Alias


m = Container()

t = Set(m, name="t", description="time in hours", records=t_conditions['Time']) # time at begining of hour 1, 2, 3, ...
p = Set(m, name="p", description="Power plant", records=p_conditions['Power plant'])

# Create alias for set p
p_up = Alias(m, name="p_up", alias_with=p)

# Parameter and Variables definitions


delay = Parameter(m, name="delay", domain=[p,p], description="Time delay for upstream plants", records=f_conditions[['From', 'To', 'Time']])

prices = Parameter(m, name="prices", domain=t, description="Prices (MWh) at different hours", records=t_conditions[['Time', 'Cost']])

reservoir_init = Parameter(m, name="reservoir_init", domain=p, description="Initial reservoir level", records=p_conditions[['Power plant', 'Initial reservoir level']])
reservoir_max = Parameter(m, name="reservoir_max", domain=p, description="Maximum reservoir level", records=p_conditions[['Power plant', 'Maximum reservoir level']])
reservoir_min = Parameter(m, name="reservoir_min", domain=p, description="Minimum reservoir level", records=p_conditions[['Power plant', 'Minimum reservoir level']])
discharge_max = Parameter(m, name="discharge_max", domain=p, description="Discharge capacity", records=p_conditions[['Power plant', 'Discharge capacity']])
power_conversion = Parameter(m, name="power_conversion", domain=p, description="Power conversion", records=p_conditions[['Power plant', 'Power conversion']])
spillage_max = Parameter(m, name="spillage_max", domain=p, description="Maximum spillage", records=p_conditions[['Power plant', 'Maximum spillage']])
local_flow = Parameter(m, name="local_flow", domain=p, description="Local flow", records=p_conditions[['Power plant', 'Local flow']])

discharge = Variable(m, name="discharge", type="positive", domain=[t,p], description="Discharge rate at each power plant at each time")
spillage = Variable(m, name="spillage", type="positive", domain=[t,p], description="Spillage rate at each power plant at each time")
reservoir_level = Variable(m, name="reservoir_level", domain=[t,p], description="Reservoir level at each power plant at each time")

### Equations and condtions

In [116]:

# Discharge criteria
discharge.up[t,p] = discharge_max[p]
discharge.up[t,p].where[(t.ord==6) & (p.ord==3)] = discharge_max[p]+1

# Spillage criteria
spillage.up[t,p] = spillage_max[p]

# Reservoir level criteria
reservoir_level.lo[t,p].where[t.ord > 1] = reservoir_min[p]
reservoir_level.up[t,p].where[t.ord > 1] = reservoir_max[p]
reservoir_level.fx[t,p].where[t.first] = reservoir_init[p] # Initial reservoir level should be set to reservoir initial level

# Single reservoir equation for all plants
reservoirs = Equation(m, name="reservoirs", domain=[t,p], description="Reservoir level at power plant (p) at different hours (t)")
reservoirs[t,p].where[t.ord > 1] = reservoir_level[t,p] == reservoir_level[t.lag(1),p] + 3600 * (
        # Upstream inflows
        Sum(p_up.where[delay[p_up,p]>0], 
            discharge[t.lag(delay[p_up, p]), p_up] + spillage[t.lag(delay[p_up, p]), p_up]
        )

        # Local inflow
        + local_flow[p]

        # Outflows
        - discharge[t.lag(1),p]
        - spillage[t.lag(1),p]
    )

### Obejctive

In [117]:
obj = Sum((t, p), prices[t]*power_conversion[p]*discharge[t,p])

## Solution

In [118]:
flow = Model(m, name="flow", equations=m.getEquations(), objective=obj, problem="LP", sense=Sense.MAX)
flow.solve()

Solver Status   Model Status         Objective Num of Equations  \
0        Normal  OptimalGlobal  334617.888888819               45   

  Num of Variables Model Type Solver Solver Time  
0              137         LP  CPLEX       0.002

In [119]:
from IPython.display import HTML

def horizontal(dfs):
    html = '<div style="display:flex">'
    for df in dfs:
        html += '<div style="margin-right: 32px">'
        html += df.to_html()
        html += '</div>'
    html += '</div>'
    display(HTML(html))

### Primals Discharge, Spillage, Reservoir Level

In [120]:
# Convert levels to pandas DataFrames for easier handling
discharge_primal_df = pd.DataFrame(discharge.l.records).pivot(index='t', columns='p', values='level')
spillage_primal_df = pd.DataFrame(spillage.l.records).pivot(index='t', columns='p', values='level')
reservoir_level_primal_df = pd.DataFrame(reservoir_level.l.records).pivot(index='t', columns='p', values='level')

horizontal([discharge_primal_df, spillage_primal_df, reservoir_level_primal_df])

p	Ahlen	Fjället	Forsen	Kärret  OBJ 334617.888888819
t				
1	0.0	    0.0	    0.000000	29.000000
2	0.0	    0.0	    0.000000	521.222222
3	0.0	    0.0	    883.777778	680.000000
4	0.0	    0.0	    0.000000	680.000000
5	522.0	33.0	975.000000	680.000000
6	540.0	135.0	1000.000000	680.000000
7	0.0	    0.0	    246.000000	680.000000
8	0.0	    0.0	    0.000000	680.000000
9	531.0	84.0	975.000000	680.000000
10	0.0	    0.0	    0.000000	0.000000
11	0.0	    56.0	271.000000	633.000000
12	0.0	    0.0	    0.000000	0.000000

p	Ahlen	Fjället	Forsen	Kärret  OBJ 334219.138888819
t				
1	0.0	    0.0	    0.000000	29.000000
2	0.0     0.0	    0.000000	521.222222
3	0.0	    0.0	    908.777778	680.000000
4	0.0	    0.0	    0.000000	680.000000
5	522.0	33.0	975.000000	680.000000
6	540.0	135.0	975.000000	680.000000
7	0.0	    0.0	    246.000000	680.000000
8	0.0	    0.0	    0.000000	680.000000
9	531.0	84.0	975.000000	680.000000
10	0.0	    0.0	    0.000000	0.000000
11	0.0	    56.0	271.000000	633.000000
12	0.0	    0.0	    0.000000	0.000000

### Duals Discharge, Spillage, Reservoir Level

In [121]:
# Convert marginals to pandas DataFrames for easier handling
discharge_duals_df = pd.DataFrame(discharge.m.records).pivot(index='t', columns='p', values='marginal')
spillage_duals_df = pd.DataFrame(spillage.m.records).pivot(index='t', columns='p', values='marginal')
reservoir_level_duals_df = pd.DataFrame(reservoir_level.m.records).pivot(index='t', columns='p', values='marginal')

horizontal([discharge_duals_df, spillage_duals_df, reservoir_level_duals_df])

